contracts/Joint.sol

```diff
@@ -159,85 +185,302 @@ contract Joint {
+        _bBalance = _bBalance.add(balanceOfB());
+    }
+
+    function estimatedTotalAssetsInToken(address token)
+        external
+        view
+        returns (uint256 _balance)
+    {
+        require(token == tokenA || token == tokenB);
+        if (token == tokenA) {
+            (_balance, ) = estimatedTotalAssetsAfterBalance();
         } else {
-            distributeProfit();
+            (, _balance) = estimatedTotalAssetsAfterBalance();
         }
     }
 
-    function createLP() internal {
-        IUniswapV2Router02(router).addLiquidity(
-            tokenA,
-            tokenB,
-            balanceOfA(),
-            balanceOfB(),
-            0,
-            0,
-            address(this),
-            now
-        );
+    event Ratios(uint256 tokenA, uint256 tokenB, string description);
+    event SellToBalance(
+        address sellToken,
+        uint256 sellAmount,
+        uint256 buyAmount
+    );
+
+    function calculateSellToBalance(
+        uint256 currentA,
+        uint256 currentB,
+        uint256 startingA,
+        uint256 startingB
+    ) internal view returns (address _sellToken, uint256 _sellAmount) {
+        if (startingA == 0 || startingB == 0) return (address(0), 0);
+
+        (uint256 ratioA, uint256 ratioB) =
+            getRatios(currentA, currentB, startingA, startingB);
+
+        if (ratioA == ratioB) return (address(0), 0);
+
+        uint256 numerator;
+        uint256 denominator;
+        uint256 precision;
+        uint256 exchangeRate;
+        if (ratioA > ratioB) {
+            _sellToken = tokenA;
+            precision = 10**uint256(IERC20Extended(tokenA).decimals());
+            (, exchangeRate) = getSpotExchangeRates();
+            numerator = currentA.sub(startingA.mul(currentB).div(startingB));
+            denominator =
+                precision +
+                startingA.mul(exchangeRate).div(startingB);
+        } else {
+            _sellToken = tokenB;
+            precision = 10**uint256(IERC20Extended(tokenB).decimals());
+            (exchangeRate, ) = getSpotExchangeRates();
+            numerator = currentB.sub(startingB.mul(currentA).div(startingA));
+            denominator =
+                precision +
+                startingB.mul(exchangeRate).div(startingA);
+        }
+        _sellAmount = numerator.mul(precision).div(denominator);
     }
 
-    function setMasterChef(address _masterchef) external onlyGov {
-        masterchef = IMasterchef(_masterchef);
-        IERC20(getPair()).approve(_masterchef, type(uint256).max);
+    function getRatios(
+        uint256 currentA,
+        uint256 currentB,
+        uint256 startingA,
+        uint256 startingB
+    ) internal pure returns (uint256 _a, uint256 _b) {
+        _a = currentA.mul(RATIO_PRECISION).div(startingA);
+        _b = currentB.mul(RATIO_PRECISION).div(startingB);
     }
 
-    function setPid(uint256 _newPid) external onlyGov {
-        pid = _newPid;
+    function getSpotExchangeRates()
```

Comment:
warning: 

this function returns a scaled number (with more decimals than real quantity)

no consequences in this implementation but might be worth it to signal in name or leave a NOTE for the future